### ISSUE TITLE
Make ref model CNN architectures auto-adaptive to number of input features

### ISSUE THREAD

**Issue by dcornet** [GBeurier/nirs4all#15]
I’d like to suggest we update our CNN architectures (decon, nicon, etc.) to make them dynamically adapt their depth and kernel sizes to the number of input features (e.g., spectral channels). Right now, our models are tuned for around 1050 features, but in practice we often see datasets from other instruments with more (or fewer) variables.

The proposed model dynamically adjusts its convolutional blocks, kernel sizes, and dense layers based on the input feature size, making it robust across different instruments or spectral configurations. If your data contains a significantly different number of features, this adaptation ensures a suitable network depth and complexity.

**For DECON**
```
from tensorflow.keras import Sequential
from tensorflow.keras.layers import (Input, SpatialDropout1D, DepthwiseConv1D, 
                                     MaxPooling1D, BatchNormalization, 
                                     SeparableConv1D, Conv1D, Flatten, Dense, Dropout)

@framework('tensorflow')
def deconAA(input_shape, params={}):
    n_features = input_shape[0]
    n_blocks = max(1, int(n_features / 500))
    min_kernel = 3
    max_kernel = 31
    min_pool = 2
    max_pool = 8
    dense_units = max(8, int(n_features * 0.07))

    model = Sequential()
    model.add(Input(shape=input_shape))
    model.add(SpatialDropout1D(params.get('spatial_dropout', 0.1)))

    current_len = n_features
    for i in range(n_blocks):
        kernel_frac = max(0.005, 0.012 * (0.8 ** i))
        pool_frac = max(0.10, 0.25 * (0.8 ** i))
        kernel_size = int(np.clip(int(current_len * kernel_frac), min_kernel, min(current_len, max_kernel)))
        pool_size = int(np.clip(int(current_len * pool_frac), min_pool, min(current_len, max_pool)))

        # Depthwise convolution
        model.add(DepthwiseConv1D(
            kernel_size=kernel_size, padding="same", depth_multiplier=2, activation="relu"
        ))

        # Only add pooling if pool_size is at least 2 and <= current_len
        if pool_size >= 2 and current_len >= pool_size:
            model.add(MaxPooling1D(pool_size=pool_size, strides=pool_size))
            current_len = int(np.ceil(current_len / pool_size))
        else:
            print(f"Skipping pooling at block {i+1}: current_len={current_len}, pool_size={pool_size}")
            # Optional: Update current_len if you used stride > 1 in conv above (here, stride=1 so nothing to do)
            pass

        model.add(BatchNormalization())

        if current_len < 3:
            print(f"⚠️ Warning: Feature map length dropped below 3 at block {i+1}. Consider reducing number of blocks or pool fraction.")
            break

    if current_len < 3:
        print(f"⚠️ Warning: Final feature map length is {current_len}. You may want to reduce blocks, kernel or pool fractions.")

    model.add(Flatten())
    model.add(Dense(dense_units, activation="relu"))
    model.add(Dense(1, activation="sigmoid"))
    return model
```

**For NICON:**
This model automatically adapts its convolutional strides, kernel sizes, and number of filters based on the input feature dimension.
Strides are adjusted progressively to ensure that the final feature map is never too small (minimum length ≥ 3), even for short input sequences. The stride now increases progressively with the input length. This ensures:
- Small inputs: The stride stays at or near 1, so we avoid losing valuable information by skipping too many features.
- Large inputs: The stride grows smoothly, making the network more efficient and preventing the feature maps from being excessively long and computationally heavy.
This flexible design improves robustness and generalization across datasets of varying sizes and avoids information loss during feature extraction.

```
from tensorflow.keras import Sequential
from tensorflow.keras.layers import (Input, SpatialDropout1D, Conv1D, 
                                     Dropout, BatchNormalization, Flatten, Dense)
import numpy as np

def get_filters(n_features, base=8, factor=0.01, max_filters=128):
    return min(int(base + n_features * factor), max_filters)

@framework('tensorflow')
def nicon(input_shape, params={}):
    n_features = input_shape[0]
    n_blocks = 3  # or adapt as you want
    min_kernel = 3
    max_kernel = 31
    min_pool = 2
    max_pool = 8
    dense_units = max(8, int(n_features * 0.015))

    model = Sequential()
    model.add(Input(shape=input_shape))
    model.add(SpatialDropout1D(params.get('spatial_dropout', 0.08)))

    current_len = n_features
    filters_list = [8, 32, 16]  # adapt or make dynamic as needed
    activations = ["selu", "relu", "elu"]

    for i in range(n_blocks):
        kernel_frac = max(0.005, 0.012 * (0.8 ** i))
        pool_frac = max(0.10, 0.25 * (0.8 ** i))
        kernel_size = int(np.clip(int(current_len * kernel_frac), min_kernel, min(current_len, max_kernel)))
        pool_size = int(np.clip(int(current_len * pool_frac), min_pool, min(current_len, max_pool)))

        # Conv layer
        model.add(Conv1D(filters=filters_list[i], kernel_size=kernel_size, strides=1,
                         activation=activations[i], padding="same"))
        if i == 0:
            model.add(Dropout(params.get('dropout_rate', 0.2)))
        model.add(BatchNormalization())

        # Pooling only if pool_size <= current_len and pool_size >= 2
        if pool_size >= 2 and current_len >= pool_size:
            model.add(MaxPooling1D(pool_size=pool_size, strides=pool_size))
            current_len = int(np.ceil(current_len / pool_size))
        else:
            print(f"Skipping pooling at block {i+1}: current_len={current_len}, pool_size={pool_size}")
            # current_len stays the same or can be updated based on stride if you use it in conv

        if current_len < 3:
            print(f"⚠️ Warning: Feature map length dropped below 3 at block {i+1}. Consider reducing number of blocks or pool fraction.")
            break

    if current_len < 3:
        print(f"⚠️ Warning: Final feature map length is {current_len}. You may want to reduce blocks, kernel or pool fractions.")

    model.add(Flatten())
    model.add(Dense(dense_units, activation="sigmoid"))
    model.add(Dense(1, activation="sigmoid"))
    return model

```
